release-2.1: changefeedccl: support TIMESTAMP and DECIMAL with the avro format

danhhz opened the pull request:
Backport 1/1 commits from #31104.

/cc @cockroachdb/release

---

Timestamp is relatively straightforward and uses the corresponding
logical type.

The avro spec also has a logical type for decimals, but ours cover a
number of cases that it doesn't (NaN, Inf, per-datum precision and
scale). I think the right answer for them is to make this a union with a
`crdb.decimal` logical type that has a string representation.
Unfortunately, we can't do that without some surgery to the avro library
we use and it seems too late in the game to do that. For now, leave them
unsupported.

A number of column types are still unsupported with avro, but this is
probably a good enough core that we can mention the experimental
support.

Release note (enterprise change): CHANGEFEEDs now have limited and
experimental support for the avro format.

danhhz committed:
changefeedccl: support TIMESTAMP and DECIMAL with the avro format

Timestamp is relatively straightforward and uses the corresponding
logical type.

The avro spec also has a logical type for decimals, but ours cover a
number of cases that it doesn't (NaN, Inf, per-datum precision and
scale). I think the right answer for them is to make this a union with a
`crdb.decimal` logical type that has a string representation.
Unfortunately, we can't do that without some surgery to the avro library
we use and it seems too late in the game to do that. For now, leave them
unsupported.

A number of column types are still unsupported with avro, but this is
probably a good enough core that we can mention the experimental
support.

Release note (enterprise change): CHANGEFEEDs now have limited and
experimental support for the avro format.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31143)
<!-- Reviewable:end -->

danhhz committed:
changefeedccl: fix TestAvroSchema/DECIMAL flake

The precision is really meant to be in [1,10], but it sure looks like
there's an off by one error in the avro library that makes this test
flake if it picks precision of 1.

Release note: None

danhhz commented:
I considered replacing this with a new "Backport 2/2 commits" PR but decided I feel like the flaky test fix is small enough to just sneak in
